src/entry.c

```diff
@@ -0,0 +1,363 @@
+}
+
+/* Modify the expiration time of this entry and return a pointer to the (potentially new) entry. */
+entry *entrySetExpiry(entry *e, long long expiry) {
+    if (entryHasExpiry(e)) {
+        char *buf = sdsAllocPtr(e);
+        if (entryHasValuePtr(e)) buf -= sizeof(sds *);
+        buf -= sizeof(expiry);
+        memcpy(buf, &expiry, sizeof(expiry));
+        return e;
+    }
+    entry *new_entry = entryUpdate(e, NULL, expiry);
+    return new_entry;
+}
+
+/* Return true in case the entry has assigned expiration or false otherwise. */
+int entryIsExpired(entry *entry) {
+    /* Don't expire anything while loading. It will be done later. */
+    if (server.loading) return 0;
+    if (!timestampIsExpired(entryGetExpiry(entry))) return 0;
+    if (server.primary_host == NULL && server.import_mode) {
+        if (server.current_client && server.current_client->flag.import_source) return 0;
+    }
+    return 1;
+}
+/**************************************** Entry Expiry API - End *****************************************/
+
+void entryFree(entry *entry) {
+    if (entryHasValuePtr(entry)) {
+        sdsfree(*entryGetValueRef(entry));
+    }
+    zfree(entryAllocPtr(entry));
+}
+
+/* Takes ownership of value. does not take ownership of field */
+entry *entryCreate(sds field, sds value, long long expiry) {
+    sds embedded_field_sds;
+    size_t expiry_size = (expiry == EXPIRY_NONE) ? 0 : sizeof(long long);
+    size_t field_len = sdslen(field);
+    int field_sds_type = sdsReqType(field_len);
+    if (field_sds_type == SDS_TYPE_5 && (expiry_size > 0)) {
+        field_sds_type = SDS_TYPE_8;
+    }
+    size_t field_size = sdsReqSize(field_len, field_sds_type);
+    size_t value_len = value ? sdslen(value) : 0;
+    size_t embedded_value_size = value ? sdsReqSize(value_len, SDS_TYPE_8) : 0;
+    size_t alloc_size = field_size + expiry_size;
+    bool embed_value = false;
+    if (value) {
+        if (alloc_size + embedded_value_size <= EMBED_VALUE_MAX_ALLOC_SIZE) {
+            /* Embed field and value. Value is fixed to SDS_TYPE_8. Unused
+             * allocation space is recorded in the embedded value's SDS header.
+             *
+             *     +------+--------------+---------------+
+             *     | TTL  | field        | value         |
+             *     |      | hdr "foo" \0 | hdr8 "bar" \0 |
+             *     +------+--------------+---------------+
+             */
+            embed_value = true;
+            alloc_size += embedded_value_size;
+        } else {
+            /* Embed field, but not value. Field must be >= SDS_TYPE_8 to encode to
+             * indicate this type of entry.
+             *
+             *     +------+-------+---------------+
+             *     | TTL  | value | field         |
+             *     |      | ptr   | hdr8 "foo" \0 |
+             *     +------+-------+---------------+
+             */
+            embed_value = false;
+            alloc_size += sizeof(sds *);
```

Comment:
an `sds` is a little unusual in that it IS a pointer.  So an `sds*` is a `char**`.  So I'm thinking you just want `sizeof(sds)`.  (Of course they're the same, but I think `sizeof(sds)` is more logically correct.)